remix-analyzer/src/solidity-analyzer/modules/forLoopIteratesOverDynamicArray.js

```diff
@@ -8,9 +8,8 @@ function forLoopIteratesOverDynamicArray () {
 }
 
 forLoopIteratesOverDynamicArray.prototype.visit = function (node) {
-  if (common.isForLoop(node) &&
-    node.children[1].children[1].attributes.member_name === 'length' &&
-    node.children[1].children[1].children[0].attributes.type.indexOf('[]') !== -1) {
+  if (common.isForLoop(node) && (common.isDynamicArrayLengthAccess(node.children[1].children[1]) ||
+  (node.children[1].children[1].children && common.isDynamicArrayLengthAccess(node.children[1].children[1].children[0])))) {
```

Comment:
It's really difficult to understand what this if statement is doing. Could you break it into several steps with explcit naming : 
```javascript
const XXX = node.children[1].children[1].children
const XXXchild = node.children[1].children[1].children[0]
const isXXX = !!XXX && common.isDynamicArrayLengthAccess(XXXchild)
const isYYY = isXXX || common.isDynamicArrayLengthAccess(node.children[1].children[1])
const isZZZ = common.isForLoop(node) && isYYY
if (common.isForLoop(node) && isZZZ) ...
```

Reply:
Also for better readibility you may want to destructure the "common" object : 
```javascript
const { isForLoop, isDynamicArrayLengthAccess } = require('./staticAnalysisCommon')
```

Reply:
sure

Reply:
I tried above suggestion of dividing checks but that was requiring hierarchical checks for existence otherwise tests were not working like 
`const XXX = node && node.children && node.children[1] && node.children[1].children[1].children`

I used a way which solves our purpose of code explanation using comments